client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -82,9 +91,39 @@ public QueryStats getStats()
     public void close()
             throws SQLException
     {
-        if (closed.compareAndSet(false, true)) {
-            ((AsyncIterator<?>) results).cancel();
-            client.close();
+        try (Closer closer = Closer.create()) {
+            synchronized (this) {
+                if (closed) {
+                    // Already closed
+                    return;
+                }
+                closed = true;
+
+                for (Runnable runnable : Lists.reverse(afterClose)) {
+                    closer.register(runnable::run);
+                }
+                afterClose.clear();
+            }
+            closer.register(client);
+            closer.register(((AsyncIterator<?>) results)::cancel);
+        }
+        catch (IOException | RuntimeException e) {
+            throw new SQLException("Close failed", e);
+        }
+    }
+
+    void runAfterClose(Runnable afterClose)
+    {
+        requireNonNull(afterClose, "afterClose is null");
+        boolean closed;
+        synchronized (this) {
+            closed = this.closed;
+            if (!closed) {
+                this.afterClose.add(afterClose);
```

Comment:
Let's rename either the field or the parameter. Having them shadow makes the code harder to read.